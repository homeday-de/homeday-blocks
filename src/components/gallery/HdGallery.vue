<template>
  <div
    :class="{
      hasSingleItem,
    }"
    class="gallery"
  >
    <figure
      v-if="hasImages"
      class="gallery__figure"
    >
      <figcaption
        v-if="showCaption"
        class="gallery__caption">{{ currentItemCaption || '&#8203;' }}</figcaption>
      <div class="gallery__main">
        <HdGalleryMedia
          :item="currentItem"
          :aspect-ratio="aspectRatio"
        />
        <div
          v-if="items.length > 1"
          class="gallery__controls"
        >
          <button
            :class="{
              'gallery__controls__prev': true,
              'isDisabled': isFirstItem,
            }"
            type="button"
            @click="goPrev"
          />
          <button
            :class="{
              'gallery__controls__next': true,
              'isDisabled': isLastItem,
            }"
            type="button"
            @click="goNext"
          />
        </div>
        <p class="gallery__info">{{ `${currentItemIndex + 1}/${items.length}` }}</p>
      </div>
    </figure>
    <HdGalleryPlaceholder
      v-else
      :aspect-ratio="aspectRatio"
      :icon="placeholderIcon"
      :text="placeholderText"
    />
    <slot name="middle"/>
    <HdGalleryCarousel
      v-if="hasImages"
      :items="items"
      v-model="currentItemIndex"
      :pager-inside="pagerInside"
      :aspect-ratio="aspectRatio"
      :disable-key-events="disableKeyEvents"
      class="gallery__carousel"
      @itemClick="onCarouselItemClick"
    />
  </div>
</template>

<script>
import HdGalleryCarousel from 'hd-blocks/components/gallery/HdGalleryCarousel.vue';
import HdGalleryMedia from 'hd-blocks/components/gallery/HdGalleryMedia.vue';
import HdGalleryPlaceholder from 'hd-blocks/components/gallery/HdGalleryPlaceholder.vue';

export default {
  name: 'HdGallery',
  components: {
    HdGalleryCarousel,
    HdGalleryMedia,
    HdGalleryPlaceholder,
  },
  props: {
    items: {
      type: Array,
      default: () => [],
    },
    pagerInside: {
      type: Boolean,
      default: false,
    },
    aspectRatio: {
      type: Number,
      default: 16 / 9,
    },
    disableKeyEvents: {
      type: Boolean,
      default: false,
    },
    startIndex: {
      type: Number,
      default: 0,
    },
    showCaption: {
      type: Boolean,
      default: true,
    },
    placeholderIcon: {
      type: String,
      // eslint-disable-next-line global-require
      default: require('hd-blocks/assets/icons/ic_photos.svg'),
    },
    placeholderText: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      currentItemIndex: this.startIndex,
    };
  },
  computed: {
    currentItem() {
      return this.items[this.currentItemIndex] || this.items[0];
    },
    currentItemCaption() {
      return this.currentItem.caption;
    },
    hasImages() {
      return this.items.length > 0;
    },
    hasSingleItem() {
      return this.items.length === 1;
    },
    isFirstItem() {
      return this.currentItemIndex === 0;
    },
    isLastItem() {
      return this.currentItemIndex === this.items.length - 1;
    },
  },
  watch: {
    currentItem(currentItem, oldCurrentItem) {
      if (currentItem.image === oldCurrentItem.image) {
        return;
      }

      this.$emit('input', currentItem);
    },
  },
  methods: {
    goPrev() {
      if (this.isFirstItem) {
        return;
      }

      this.currentItemIndex -= 1;
    },
    goNext() {
      if (this.isLastItem) {
        return;
      }

      this.currentItemIndex += 1;
    },
    onCarouselItemClick(itemIndex) {
      this.$emit('carouselItemClick', itemIndex);
    },
  },
};
</script>

<style lang="scss">
@import 'hd-blocks/styles/mixins.scss';

.gallery {
  $_root: &;
  position: relative;
  pointer-events: none;

  > * {
    pointer-events: all;
  }

  &__figure {
    pointer-events: none;

    > * {
      pointer-events: all;
    }
  }

  &__caption {
    margin-bottom: $stack-s;
  }

  &__main {
    position: relative;
    display: none;

    @media (min-width: $break-tablet) {
      display: block;
    }

    &__info {
      position: absolute;
      right: $inline-xs;
      bottom: $stack-xs;
      padding-right: $inline-s;
      padding-left: $inline-s;
      background-color: rgba(0, 0, 0, 0.8);
      @include font('text-xsmall');
      font-weight: 600;
      color: white;
      border-radius: 2px;
    }
  }

  &__controls {
    display: none;

    @media (min-width: $break-tablet) {
      display: block;
    }

    &__prev,
    &__next {
      display: flex;
      width: 36px;
      height: 36px;
      padding: 0;
      position: absolute;
      top: 50%;
      left: 0;
      transform: translate(-30%, -50%);
      align-items: center;
      justify-content: center;
      appearance: none;
      background-color: $white;
      border-radius: 50%;
      border: 1px solid $ebb2;
      box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
      cursor: pointer;
      transition: opacity $time-s ease-in-out;

      @media (min-width: $break-desktop) {
        transform: translate(-50%, -50%);
      }

      &.isDisabled {
        opacity: 0;
        pointer-events: none;
      }

      &::before {
        content: '';
        display: block;
        height: 24px;
        width: 24px;
        background-image: url('~hd-blocks/assets/icons/ic_arrow.svg');
        background-position: center;
        background-size: contain;
        background-repeat: no-repeat;
        transform: rotateY(180deg);
      }
    }

    &__next {
      left: auto;
      right: 0;
      transform: translate(30%, -50%);

      @media (min-width: $break-desktop) {
        transform: translate(50%, -50%);
      }

      &::before {
        transform: none;
      }
    }
  }

  &__info {
    position: absolute;
    right: $inline-xs;
    bottom: $stack-xs;
    padding-right: $inline-s;
    padding-left: $inline-s;
    background-color: rgba(0, 0, 0, 0.8);
    @include font('text-xsmall');
    font-weight: 600;
    color: white;
    border-radius: 2px;
  }

  &__carousel {
    #{$_root}.hasSingleItem & {
      @media (min-width: $break-tablet) {
        display: none;
      }
    }
  }
}
</style>
