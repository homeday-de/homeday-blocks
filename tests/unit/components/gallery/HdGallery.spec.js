import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdGallery from '@/components/gallery/HdGallery.vue';
import ITEMS from '@/stories/mocks/GALLERY_ITEMS';

const FIGURE_SELECTOR = '.gallery__figure';

const wrapperBuilder = wrapperFactoryBuilder(HdGallery, {
  props: {
    items: ITEMS,
  },
});

describe('HdGallery', () => {
  it('renders component', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('switchs elements and emits the proper events clicks', () => {
    const wrapper = wrapperBuilder();
    const testIndex = 3;
    const figure = wrapper.find(FIGURE_SELECTOR);

    figure.trigger('click');
    expect(wrapper.emitted('currentItemClick')[0][0]).toBe(0);

    wrapper.setData({ currentItemIndex: testIndex });

    figure.trigger('click');
    expect(wrapper.emitted('currentItemClick')[1][0]).toBe(testIndex);
  });
});
