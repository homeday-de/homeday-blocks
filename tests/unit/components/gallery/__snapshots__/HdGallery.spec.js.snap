// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdGallery renders component 1`] = `
<div class="gallery">
  <figure class="gallery__figure">
    <figcaption class="gallery__caption">Image 1</figcaption>
    <div class="gallery__main">
      <div class="gallery-media">
        <div class="gallery-media__object">
          <div class="gallery-media__object__sizer" style="padding-top: 56.25%;"></div>
          <picture class="gallery-media__object__picture">
            <source media="(max-width: 600px)" srcset="https://dummyimage.com/306x172?text=1+mw+600px">
            <source media="(max-width: 900px)" srcset="https://dummyimage.com/768x432?text=1+mw+900px">
            <source media="(max-width: 1200px)" srcset="https://dummyimage.com/768x432?text=1+mw+1200px"> <img src="https://dummyimage.com/1536x364?text=1" alt="Image 1"></picture>
          <div class="gallery-media__object__thumbnail isVisible" style="background-image: url(https://dummyimage.com/154x86?text=1+thumb);"></div>
        </div>
      </div>
      <div class="gallery__controls"><button type="button" class="gallery__controls__prev isDisabled">
          <!----></button> <button type="button" class="gallery__controls__next">
          <!----></button></div>
      <p class="gallery__info">1/10</p>
    </div>
  </figure>
  <div class="gallery__carousel gallery-carousel">
    <div tabindex="0" class="gallery-carousel__wrap">
      <div class="flickity-enabled">
        <div class="flickity-viewport" style="height: 0px;">
          <div class="flickity-slider" style="left: 0px; transform: translateX(NaN%);">
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture">
                <source media="(max-width: 900px)" srcset="https://dummyimage.com/768x432?text=1+thumb+mw+900px"> <img src="https://dummyimage.com/154x86?text=1+thumb" alt="Image 1"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=2+thumb" alt="Image 2" srcset="https://dummyimage.com/308x172?text=2+thumb+DPR2x 2x, https://dummyimage.com/462x258?text=2+thumb+DPR3x 3x"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=3+thumb" alt="Image 3"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=4+thumb" alt="Image 4"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=5+thumb" alt="Image 5"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=6+thumb" alt="Image 6"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=7+thumb" alt="Image 7"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=8+thumb" alt="Image 8"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=9+thumb" alt="Image 9"></picture>
            </div>
            <div class="gallery-carousel__item is-selected" style="position: absolute;">
              <div class="gallery-carousel__item__sizer" style="padding-top: 56.25%;"></div>
              <picture class="gallery-carousel__item__picture"> <img src="https://dummyimage.com/154x86?text=10+thumb" alt="Image 10"></picture>
            </div>
          </div>
        </div>
      </div>
      <div class="gallery-carousel__pager">
        <!---->
      </div>
    </div>
  </div>
</div>
`;
